ln-gateway/src/ln.rs

```diff
@@ -21,50 +21,16 @@ pub trait LnRpc: Send + Sync + 'static {
 pub struct LightningError(pub Option<i32>);
 
 #[async_trait]
-impl LnRpc for Mutex<cln_rpc::ClnRpc> {
-    #[instrument(name = "LnRpc::pay", skip(self))]
-    async fn pay(
+pub trait LnRpcFactory: Send + Sync + 'static {
+    async fn create(
         &self,
-        invoice: &str,
-        max_delay: u64,
-        max_fee_percent: f64,
-    ) -> Result<Preimage, LightningError> {
-        debug!("Attempting to pay invoice");
-
-        let pay_result = self
-            .lock()
-            .await
-            .call(cln_rpc::Request::Pay(PayRequest {
-                bolt11: invoice.to_string(),
-                amount_msat: None,
-                label: None,
-                riskfactor: None,
-                maxfeepercent: Some(max_fee_percent),
-                retry_for: None,
-                maxdelay: Some(max_delay as u16),
-                exemptfee: None,
-                localofferid: None,
-                exclude: None,
-                maxfee: None,
-                description: None,
-            }))
-            .await;
+        messenger: GatewayMessageChannel,
+    ) -> Result<Arc<LnRpcRef>, anyhow::Error>;
+}
```

Comment:
Do we really need a factory here? Couldn't this just be a `new` fn of the respective `LnRpc` impl or even part of the `LnRpc` trait (using `where Self: Sized` to preserve object safety)?

Reply:
I'm not sure different impls will have a standard `new` fn signature. Certainly though, the minimum we need to give the impl is a messaging channel so they they can 'talk' back to us once instantiated. A factory nicely abstracts away the details of the impl

Reply:
I'm not sure different impls will have a standard `new` fn signature. Certainly though, the minimum we need to give the impl is a messaging channel so they they can 'talk' back to us once instantiated. A factory nicely abstracts away the details of the impl

Reply:
> I'm not sure different impls will have a standard `new` fn signature.

I agree, so maybe just leave it at custom `new` methods? But I also don't see how different constructors would be supported using the factory. If different config is needed based on the impl that could be represented as an enum and based on the variant the respective constructor is called.

Reply:
> RE:  But I also don't see how different constructors would be supported using the factory. If different config is needed based on the impl that could be represented as an enum and based on the variant the respective constructor is called.

This would be necessary if `LnGateway` lib that calls the factory, had knowledge and preference of the LN implementation. I actually inverted the control here such that the binary uses `LnGateway` from lib, and it's the binary that selects an Ln implementation to use. See `ln_gateway.rs` which currently selects cln implementation.

To extend this so we can support multiple Ln implementations, we could either:
1. use args on a single `ln_gateway` binary to select the right LN implementation function. OR
2. build multiple `ln_gateway` binaries , each for a specific LN implementation?

the factory signature on `LnGateway::register_ln_rpc` remains simple and standard as an API. It has no knowledge of the LN impl we will use the API to inject. However, from the binary side, we could totally could augment the params needed to instantiate the LN implementation. Example

```rs
async fn main() -> Result<(), Error> {
   // ... elided: args and env parsing

   /// Augment cln implementation function with a logger param?
    // we do this here to simulate a scenario in which cln rpc needs a logger,
    // but LnGateway only provides a messenger
    fn create_cln_rpc(messenger: GatewayMessageChannel) -> Result<LnRpcRef, Error> {
        let logger = Logger::root();
        build_cln_rpc(messenger, logger)
    }

    let mut gateway = LnGateway::new();

    let ln_ref = gateway
        .register_ln_rpc(create_cln_rpc)
        .await
        .expect("Failed to register cln rpc");

    // ... elided: build and register gateway client

    gateway.run().await.expect("gateway failed to run");
    Ok(())
}
```

Reply:
Now, I took another look at my factory implementation that started this discussion... and got even more convinced that we want to keep a simple factory here. Passing `cln_rpc::ClnRpc::new()` to `register_ln_rpc` API couples CLN impl details with the `LnGateway`. Further, current cln rpc requires building of the plugin before calling `ClnRpc::new` so if I'm to avoid writing a wrapper class which takes takes the shape of a **class factory** as we were discussing above, I rewrote the factory as a function with the shape below...

```rs
pub async fn register_ln_rpc<F, R>(
        &mut self,
        create_ln_rpc: F,
    ) -> std::result::Result<LnRpcRef, anyhow::Error>
    where
        F: Fn(GatewayMessageChannel) -> R,
        R: Future<Output = std::result::Result<LnRpcRef, anyhow::Error>>,
    {
        let (sender, receiver): (mpsc::Sender<GatewayRequest>, mpsc::Receiver<GatewayRequest>) =
            mpsc::channel(100);

        // Register the lightning RPC client
        let ln_ref = create_ln_rpc(GatewayMessageChannel::new(&sender)).await?;
        
        // ... elided
   }
```

with

```rs
   pub async fn build_cln_rpc(messenger: GatewayMessageChannel) -> Result<LnRpcRef, Error> {
       // build plugin
       // locate socket
       // instantiate ln_rpc using `ClnRpc::new(socket)`
       // get pub key
       // return a reference
   }
```

OFC this is still the factory pattern, maybe just simpler and more 'rusty' to write as a function :)

Reply:
Actually it's much harder to rework our integration test framework using a factory function pattern. The lightnig tests are structs that are easier to augment with factory trait than plain function. Same outcome. If my premise for using a factory is agreeable, then I'd mark this as **"wont-fix"** :)
![image](https://user-images.githubusercontent.com/11217077/194939549-4b20301c-b69c-467d-aa83-12e32e1e8cdd.png)